{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "19f1c6b8-59cb-47b6-9845-be423dd4b230",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## Summary"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7ae80393-4528-42e8-8236-85d68b550714",
   "metadata": {
    "tags": []
   },
   "source": [
    "**Goal: Calculate peptide distribution and protein distribution P(Zj)** from P(Xi|Zj) fluorosequencing scores \n",
    "\n",
    "___\n",
    "\n",
    "**Prior to my code**\n",
    "\n",
    "1. **Acquiring of experimental data (reads)**\n",
    "    - **4-5 fluorescence level plots vs Edman degradation cycle, 1 read for each molecule**\n",
    "\n",
    "\n",
    "2. **Matts classifier converts reads into P(Xi|Zj)**, the likelihood of each partial sequence given a peptide.\n",
    "    - **Calculating P(Xi) directly impossible/difficult**. You may think that it is simply number of partial sequences/number of all reads, but that is only true if the number of reads is much larger than the number of possible outcomes. It's like rolling a die a couple times, and then asking how likely each number is. You can only do that if you are certain that your number of rolls adequately samples the possibilities. In our case, we do not know this and thus it is instead more consistent to **calculate P(Xi|Zj) instead and to use that to calculate P(Zj), the target variable.**\n",
    "    \n",
    "___\n",
    "\n",
    "**This code**\n",
    "\n",
    "1. Calculation of peptide distribution (P(Zj)) using P(Xi|Zj) scores\n",
    "    1. **imports P(Xi|Zj) table, generated by Matt's program**\n",
    "    2. Asks user for settings\n",
    "    3. For each bootstrap run:\n",
    "        1. Create a subtable created from randomly sampling rows in P(Xi|Zj)\n",
    "        2. Use this subtable of P(Xi|Zj) to run EM until P(Zj) converges. P(zj) calculation is like Monte Carlo integration, because P(Zj|Xi)  \\\n",
    "           **For each EM run:**\n",
    "            \n",
    "           - **Hen, egg problem. two missing variables: P(Zj|Xi) and P(Zj). calculating either requires the other**\n",
    "           - for **1st calculation of P(Zj|Xi) approximate: assume equal distribution of all P(Zj)**\n",
    "           - **alternate between updating P(Zj|Xi) and P(Zj) until convergence**\n",
    "                \n",
    "    4. Calculate **average P(Zi) from all bootstrap runs, and calculate 95% CI**\n",
    "    \n",
    "2. Calculation of protein distribution using calculated peptide distribution\n",
    "    1. **Calculate P(Xi|Zj) from number of matches against digested proteome**\n",
    "    1. run **bootstrap-EM, like in peptide inferring step**\n",
    "        - First calculation of P(Zj|Xi) assumes equal distribution of all P(Zj)\n",
    "        - **Uses peptide distribution P(Zj) values from (1) as P(Xi), which simplifies the Bayes-derived equation.** However, the principle is the same\n",
    "        \n",
    "___\n",
    "**Open questions**\n",
    "- Is there a case in which bootstrapping improves the guess? Test dataset with many rare peptides\n",
    "- make protein inferring work\n",
    "- suggestions for data I should be showing?\n",
    "    - plot of results depending on number of EM runs, compared to argmax and weighted score\n",
    "    - show effect of bootstrap runs and see if its advantageous to use that for any dataset\n",
    "    - both for peptide and for protein inferring"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c0b9c47a-413d-48bb-aba3-421e61fab566",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Functions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "080fed80-531c-491a-a919-d15866efa117",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "### import of packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "b9704e7d-95cf-49e7-816a-9b65de641c1c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import numpy as np # better arrays than inbuilt arrays\n",
    "import matplotlib.pyplot as plt # to plot stuff\n",
    "\n",
    "import pandas as pd #for DataFrame tables\n",
    "from IPython.display import display #to display dfs more nicely\n",
    "\n",
    "import scipy.stats\n",
    "#from scipy.stats import norm\n",
    "import statistics\n",
    "\n",
    "import math\n",
    "import time #to measure runtime"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea95e8dc-bea2-4a9e-9927-a24e879c09fe",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "### Generation of random distributions from 3 Gaussians and plotting them"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "b966b932-583d-4442-bd65-6929ad4c8caa",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def create_data_from_3_gaussians(mean1, mean2, mean3, std1, std2, std3, n1, n2, n3): # Creation of random data points from multiple Gaussians ki\n",
    "    k1 = np.random.normal(mean1, std1, n1) #creates array with values created through Gaussian\n",
    "    k2 = np.random.normal(mean2, std2, n2)\n",
    "    k3 = np.random.normal(mean3, std3, n3)\n",
    "\n",
    "    return np.concatenate([k1, k2, k3])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "298a5777-5f78-470f-8ec6-1c6b31d68e81",
   "metadata": {},
   "outputs": [],
   "source": [
    "def p_xi_given_zj_from_gaussian_datapoints(kall, mean1, mean2, mean3, std1, std2, std3):\n",
    "    # Calculating p_xi_given_zj (this is what Matt is working on with the simulated data)\n",
    "\n",
    "    pdf_probability_k1 = scipy.stats.norm.pdf(kall, loc=mean1, scale=std1)\n",
    "    pdf_probability_k2 = scipy.stats.norm.pdf(kall, loc=mean2, scale=std2)\n",
    "    pdf_probability_k3 = scipy.stats.norm.pdf(kall, loc=mean3, scale=std3)\n",
    "\n",
    "    p_xi_given_zj = np.vstack((pdf_probability_k1,pdf_probability_k2,pdf_probability_k3))\n",
    "    p_xi_given_zj = np.transpose(p_xi_given_zj)\n",
    "\n",
    "    print(\"p_xi_given_zj\")\n",
    "    display(pd.DataFrame(p_xi_given_zj))\n",
    "    print(\"\\n\")\n",
    "    \n",
    "    print(\"likeliest peptide z of each datapoint x according to scipy.stats.norm.pdf (should be more accurate than EM because its dedicated to Gaussians)\")\n",
    "    print(np.argmax(p_xi_given_zj, axis=1)) # for each column of P(Xi|Zj), the most likely Peptide is returned\n",
    "    print(\"\\n\")\n",
    "    \n",
    "    return p_xi_given_zj"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "2bd649d2-aa10-4e72-8649-5f290e8bda68",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_3_gaussians_and_calculate_p_xi_given_zj(mean1, mean2, mean3, std1, std2, std3, n1, n2, n3):\n",
    "    kall = create_data_from_3_gaussians(mean1, mean2, mean3, std1, std2, std3, n1, n2, n3)\n",
    "    p_xi_given_zj = p_xi_given_zj_from_gaussian_datapoints(kall, mean1, mean2, mean3, std1, std2, std3)\n",
    "    \n",
    "    return p_xi_given_zj"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "fc2bb9da-3d95-4ca4-bf1c-343b61fbfdf4",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def plot_histograms_and_pdfs_from_gaussians(mean1, mean2, mean3, std1, std2, std3, n1, n2, n3):\n",
    "    k1 = np.random.normal(mean1, std1, n1) #creates array with values created through Gaussian\n",
    "    k2 = np.random.normal(mean2, std2, n2)\n",
    "    k3 = np.random.normal(mean3, std3, n3)\n",
    "    \n",
    "    # plotting histograms\n",
    "    nbins = 50\n",
    "    plt.hist(k1, label = \"Peptide 0\", bins=nbins, alpha=0.3, density=True, color=\"orange\") # alpha=transparency, density=True normalises to 1 \n",
    "    plt.hist(k2, label = \"Peptide 1\", bins=nbins, alpha=0.3, density=True, color=\"green\")\n",
    "    plt.hist(k3, label = \"Peptide 2\", bins=nbins, alpha=0.3, density=True, color=\"blue\")\n",
    "\n",
    "    # PDF plot\n",
    "    xmin, xmax = plt.xlim() #finds lower and upper bounds of histogram data\n",
    "    x = np.linspace(start=xmin, stop=xmax, num=100) #num is the number of returned data points - the more points, the finer the fit is plotted\n",
    "    p1 = scipy.stats.norm.pdf(x, mean1, std1)\n",
    "    p2 = scipy.stats.norm.pdf(x, mean2, std2)\n",
    "    p3 = scipy.stats.norm.pdf(x, mean3, std3)\n",
    "\n",
    "    plt.plot(x, p1, linewidth=2, color = \"orange\", label = \"Gauss function k1: mean = {:.2f}, STD = {:.2f}\".format(mean1, std1))\n",
    "    plt.plot(x, p2, linewidth=2, color = \"green\", label = \"Gauss function k2: mean = {:.2f}, STD = {:.2f}\".format(mean2, std2))\n",
    "    plt.plot(x, p3, linewidth=2, color = \"blue\", label = \"Gauss function k3: mean = {:.2f}, STD = {:.2f}\".format(mean3, std3))\n",
    "\n",
    "    plt.legend(loc='upper right')\n",
    "    plt.title(\"PDFs of dwarves and humans\")\n",
    "\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "307eddb1-7bfa-4c33-aae3-323d2cf41b86",
   "metadata": {
    "tags": []
   },
   "source": [
    "### EM    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "4f66876d-92be-4144-873b-8117a6138ce2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def update_p_zj_given_xi(p_xi_given_zj, p_zj, p_zj_given_xi):\n",
    "    ## for PEPTIDE inference\n",
    "    if inference_mode == \"peptide\" or \"protein\":\n",
    "        denominator = 0\n",
    "        for i, row in enumerate(p_xi_given_zj): # Calculating/Updating P(Zj|Xi)\n",
    "            #print(\"ROW of P(Xi|Zj):\", i)\n",
    "            for j, cell in enumerate(row):\n",
    "                #print(\"COLUMN of P(Xi|Zj):\", j)\n",
    "                numerator = cell * p_zj[j]\n",
    "                #print(\"numerator:\", numerator)\n",
    "\n",
    "                for l, cell in enumerate(p_zj):\n",
    "                    # print(\"cell\", i, l, p_xi_given_zj[i, l], end=\"\")\n",
    "                    # print(\" * zl\", p_zj[l])\n",
    "                    denominator = denominator + p_xi_given_zj[i, l] * p_zj[l]\n",
    "                \n",
    "                # print(\"p_zj_given_xi[j][i] = numerator/denominator\", numerator, \"/\", denominator)\n",
    "                #display(pd.DataFrame(p_zj_given_xi))\n",
    "                p_zj_given_xi[j][i] = numerator/denominator\n",
    "                denominator = 0                                \n",
    "        #display(pd.DataFrame(p_zj_given_xi))    \n",
    "        #print((np.argmax(p_zj_given_xi, axis=1))) # reports index of max value from each row\n",
    "\n",
    "        \n",
    "    ## for PROTEIN inference (denominator is has no loop)\n",
    "    elif inference_mode == \"not protein idk why this doesnt work\":\n",
    "        # print(\"p_xi_given_zj (subarray)\")\n",
    "        # display(pd.DataFrame(p_xi_given_zj))\n",
    "        for i, row in enumerate(p_xi_given_zj): # Calculating/Updating P(Zj|Xi)#\n",
    "            # print(\"ROW of P(Xi|Zj):\", i)\n",
    "            for j, cell in enumerate(row):\n",
    "                # print(\"COLUMN of P(Xi|Zj):\", j)\n",
    "                # print(\"p_zj_given_xi\", cell, \"*\",  p_zj[j], \"/\", p_xi[i],\"=\", cell*p_zj[j]/p_xi[i])\n",
    "                p_zj_given_xi[j][i] = cell * p_zj[j]/p_xi[i]\n",
    "\n",
    "        # display(pd.DataFrame(p_zj_given_xi))\n",
    "       # print((np.argmax(p_zj_given_xi, axis=1))) # reports index of max value from each row\n",
    "        \n",
    "    else:\n",
    "        print(\"ERROR: inference_mode was not set to peptide or protein\")\n",
    "        \n",
    "    return p_zj_given_xi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "a1e2eba0-fa71-4faf-833d-f0bd0244f989",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def update_p_zj(p_zj, p_zj_given_xi):  \n",
    "    if inference_mode == \"peptide\":\n",
    "        for j, element in enumerate(p_zj): #updating the expectation value of Zi\n",
    "            p_zj[j] = p_zj_given_xi[j].sum()/p_xi_given_zj.shape[1] # divide by number of peptides z\n",
    "\n",
    "        p_zj_sum = np.sum(p_zj, axis = 0)\n",
    "        p_zj = p_zj/p_zj_sum\n",
    "\n",
    "        # print(\"EM loop\", EM_loopcounter, \")\")\n",
    "        # display(pd.DataFrame(p_zj))\n",
    "    \n",
    "    elif inference_mode == \"protein\":\n",
    "        p_zj.fill(0) # overwrite all values of pzj with 0\n",
    "\n",
    "        print(\"p_zj_given_xi after update\")\n",
    "        display(pd.DataFrame(p_zj_given_xi))\n",
    "        \n",
    "        for j, row in enumerate(p_zj_given_xi):            \n",
    "            for i, element in enumerate(row):\n",
    "                print(\"p_zj_given_xi[j][i] * p_xi[i]\", p_zj_given_xi[j][i], p_xi[i], p_zj_given_xi[j][i] * p_xi[i])\n",
    "                p_zj[j] = p_zj[j] + (p_zj_given_xi[j][i] * p_xi[i])\n",
    "        \n",
    "        p_zj_sum = np.sum(p_zj, axis = 0)\n",
    "        p_zj = p_zj/p_zj_sum\n",
    "        print(\"p_zj (EM loop\", EM_loopcounter, \")\")\n",
    "        display(pd.DataFrame(p_zj))\n",
    "    \n",
    "    else:\n",
    "        print(\"ERROR: inference_mode was not set to peptide or protein\")\n",
    "    \n",
    "    return p_zj"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "3c5b5972-b626-454b-8422-c5be7e306557",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def EM_convergence_checker(p_zj_old, p_zj, EM_convergence_minimum):\n",
    "    difference_abs = abs(np.sum(p_zj_old[round(0.5*len(p_zj_old))::], axis=0, dtype=float) - np.sum(p_zj[round(0.5*len(p_zj_old))::], axis=0, dtype=float))\n",
    "    if difference_abs > EM_convergence_minimum:\n",
    "        return False\n",
    "    \n",
    "    else:\n",
    "        return True   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "79e08316-e78a-4de3-b949-3ba2077fd813",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def EM(p_xi_given_zj):\n",
    "    ### initialise p_zj_given_xi\n",
    "    # p_xi_given_zj = np.arange(0,102,1).reshape(34,3) # for testing EM iterating through columns and rows\n",
    "    p_zj_given_xi = np.full((p_xi_given_zj.shape[1], p_xi_given_zj.shape[0]), 0, dtype=float) #Initialisation based on array size of p_xi_given_zj -- same size, but transposed\n",
    "    \n",
    "    ### initialise p_zj\n",
    "    n = p_xi_given_zj.shape[1] # number of peptides\n",
    "    p_zj_initial = 1/n #initial approximation: all zj equally likely, to jumpstart first iteration\n",
    "\n",
    "    global p_zj\n",
    "    p_zj = np.full(n, p_zj_initial)\n",
    "    p_zj_old = np.full(n, 0)\n",
    "    \n",
    "    global EM_loopcounter \n",
    "    EM_loopcounter = 0\n",
    "    EM_convergence_checker_result = False\n",
    "        \n",
    "    while (EM_convergence_checker_result == False) and EM_loopcounter < EM_loopcounter_max:\n",
    "        p_zj_given_xi = update_p_zj_given_xi(p_xi_given_zj, p_zj, p_zj_given_xi)\n",
    "        \n",
    "        p_zj_old = np.copy(p_zj, order='K', subok=False)\n",
    "        p_zj = update_p_zj(p_zj, p_zj_given_xi)\n",
    "        EM_convergence_checker_result = EM_convergence_checker(p_zj_old, p_zj, EM_convergence_minimum)\n",
    "\n",
    "        EM_loopcounter = EM_loopcounter + 1\n",
    "    return p_zj"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c5084a15-17a8-498a-a95e-024edc219205",
   "metadata": {
    "tags": []
   },
   "source": [
    "### bootstrapping"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "61baafa7-6822-4221-b27a-83be3803ce68",
   "metadata": {},
   "source": [
    "**create_subarray_of_p_xi_given_zj(p_xi_given_zj)**\n",
    "\n",
    "   - Returns subarray of P(Xi|Zj) input data. \n",
    "   - Subarray's size is a fraction of the full datasize, this fraction is user-specified (default: 0.7). Sampling is done \"with replacement\".\n",
    "   \n",
    "**bootstrap_EM()**\n",
    "\n",
    "   - returns: P(Zj) values after bootstrap/EM, in relative values (i. e. in numbers <= 1)\n",
    "   - if bootstrap_sampled_fraction is set to -1, no bootstrapping occurs, and just one EM using full P(Xi|Zj) dataset is run\n",
    "   - else, multiple bootstraps are run (default: 200), in which each EM uses only a subarray created from the full dataset using the create_subarray_of_p_xi_given_zj function\n",
    "   \n",
    "**bootstrap_EM_analytics_AVG():**\n",
    "\n",
    "   - returns: the average P(Zj) from all bootstrap runs\n",
    "   \n",
    "**bootstrap_EM_analytics_CI(CI_percent)**\n",
    "\n",
    "   - returns: confidence interval of all bootstrap runs (default: 95% CI)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "017f6966-6cfe-47b9-b11c-46cbe649985e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def create_subarray_of_p_xi_given_zj(p_xi_given_zj):\n",
    "    df_p_xi_given_zj = pd.DataFrame(p_xi_given_zj)\n",
    "    df_p_xi_given_zj_sample = df_p_xi_given_zj.sample(frac=bootstrap_sampled_fraction, axis='rows', replace=True) # filters for a random partial dataset\n",
    "    # display(pd.DataFrame(df_p_xi_given_zj_sample))\n",
    "    p_xi_given_zj_subarray = df_p_xi_given_zj_sample.to_numpy()\n",
    "    \n",
    "    # print(\"p_xi_given_zj_subarray\")\n",
    "    # display(pd.DataFrame(p_xi_given_zj_subarray))\n",
    "    \n",
    "    return p_xi_given_zj_subarray"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "4edffbe3-83fa-4e42-bf2f-17beafeb014a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def bootstrap_EM():\n",
    "    i = 0\n",
    "    \n",
    "    p_zj_bootstrap_results_absolute = np.full((p_xi_given_zj.shape[1]), 0, dtype=float)\n",
    "    \n",
    "    if bootstrap_sampled_fraction == -1: # no bootstrapping, i. e. use full dataset for EM. Since the EM is deterministic, there is no point in running multiple bootstrap runs\n",
    "        p_xi_given_zj_subarray = p_xi_given_zj\n",
    "        p_zj_bootstrap_results_absolute = EM(p_xi_given_zj_subarray)\n",
    "    \n",
    "    else: # with bootstrapping, i. e. a subarray of the dataset is created for each bootstrap run\n",
    "        while i < n_bootstrap_runs:\n",
    "            p_xi_given_zj_subarray = create_subarray_of_p_xi_given_zj(p_xi_given_zj) \n",
    "            if i == 0:\n",
    "                p_zj_bootstrap_results_absolute = EM(p_xi_given_zj_subarray)\n",
    "            else:\n",
    "                p_zj_bootstrap_results_absolute = np.vstack((p_zj_bootstrap_results_absolute, EM(p_xi_given_zj_subarray)))\n",
    "            \n",
    "            print(\"Bootstrap run #\", i, \". EM loops: \", EM_loopcounter, sep=\"\")\n",
    "            if EM_loopcounter == EM_loopcounter_max:\n",
    "                print(\"WARNING: EM_loopcounter was reached, convergence likely still has not been reached. Consider increasing the maximum number of EM loops.\")\n",
    "\n",
    "            i = i + 1\n",
    "    \n",
    "    #print(\"P_zj_bootstrap_results_absolute\", p_zj_bootstrap_results_absolute)\n",
    "    \n",
    "    global p_zj_bootstrap_results_fraction\n",
    "    if bootstrap_sampled_fraction == -1:\n",
    "         p_zj_bootstrap_results_fraction = p_zj_bootstrap_results_absolute/np.sum(p_zj_bootstrap_results_absolute, axis = 0)\n",
    "    \n",
    "    else:\n",
    "        p_zj_bootstrap_results_fraction = p_zj_bootstrap_results_absolute/np.sum(p_zj_bootstrap_results_absolute, axis = 1)[0]\n",
    "        \n",
    "    print(\"Zj values from every bootstrap run (columns: peptides, rows: , displayed as fractions: \")\n",
    "    display(pd.DataFrame(p_zj_bootstrap_results_fraction))\n",
    "    return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "9df928ed-adcb-402d-9ec5-97ffd317a4d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def bootstrap_EM_analytics_AVG():\n",
    "    global p_zj_bootstrap_results_fraction_avg\n",
    "    \n",
    "    if bootstrap_sampled_fraction == -1:\n",
    "        p_zj_bootstrap_results_fraction_avg = np.copy(p_zj_bootstrap_results_fraction,order='K')\n",
    "    else:\n",
    "        p_zj_bootstrap_results_fraction_avg = np.sum(p_zj_bootstrap_results_fraction, axis = 0)/p_zj_bootstrap_results_fraction.shape[0]\n",
    "        \n",
    "    # print(\"AVG zj bootstrapping values, displayed as fractions: \\n\", p_zj_bootstrap_results_fraction_avg, sep=\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "ec5ce1eb-7e35-4cf1-b20a-b632273a4487",
   "metadata": {},
   "outputs": [],
   "source": [
    "def bootstrap_EM_analytics_CI():\n",
    "    highindex = round(CI_percent*len(p_zj_bootstrap_results_fraction))\n",
    "    lowindex = round((1-CI_percent)*len(p_zj_bootstrap_results_fraction))\n",
    "    \n",
    "    bootstrap_CI_max = np.sort(p_zj_bootstrap_results_fraction, axis=0, kind=\"quicksort\", order=None)[lowindex:highindex:][-1] #fetches last row from sorted table, ie the highest values. axis=0: sorting along each column\n",
    "    bootstrap_CI_min = np.sort(p_zj_bootstrap_results_fraction, axis=0, kind=\"quicksort\", order=None)[lowindex:highindex:][0] #same but for first row, highest values\n",
    "    bootstrap_CI_minmax = np.stack((bootstrap_CI_min, bootstrap_CI_max))\n",
    "    \n",
    "    ### show full sorted table\n",
    "    # global p_zj_bootstrap_results_fraction_CI\n",
    "    # print(\"The\", CI_percent*100, \"confidence interval of all zj bootstrapping values sorted, displayed as fractions: \")    \n",
    "    # p_zj_bootstrap_results_fraction_CI = np.sort(p_zj_bootstrap_results_fraction, axis=0, kind=\"quicksort\", order=None)[lowindex:highindex:] #axis=sorting along each column\n",
    "    # display(pd.DataFrame(p_zj_bootstrap_results_fraction_CI))\n",
    "    \n",
    "    return bootstrap_CI_minmax"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d8adb5e-3057-424e-866e-637ebccd86ee",
   "metadata": {
    "tags": []
   },
   "source": [
    "## call functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "c6c157ec-dcbd-457a-a448-bde516d2728d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "EM_convergence_minimum? If nothing is entered, it is set to 0.0001. \n",
      "Maximum number of EM runs (per bootstrap run)? If nothing is entered, it is set to 200. \n",
      "Fraction of subarray sampled for each bootstrap run? If nothing is entered, it is set to 0.8. If -1 is entered, bootstrapping is turned off \n",
      "Number of bootstrap runs? If nothing is entered, it is set to 200. \n",
      "Condidence interval? If nothing is entered, it is set to 95. \n"
     ]
    }
   ],
   "source": [
    "# esc, ctrl+a, ctrl+enter to run all cells\n",
    "\n",
    "#p_xi_given_zj = create_3_gaussians_and_calculate_p_xi_given_zj(140, 150, 160, 15, 15, 15, 5, 5, 5) #mean1-3, std1-3, n1-3\n",
    "# plot_histograms_and_pdfs_from_gaussians(140, 150, 160, 15, 15, 15, 5, 5, 5)\n",
    "\n",
    "EM_convergence_minimum = float(input(\"EM_convergence_minimum? If nothing is entered, it is set to 0.0001.\") or \"0.0001\")\n",
    "EM_loopcounter_max = int(input(\"Maximum number of EM runs (per bootstrap run)? If nothing is entered, it is set to 200.\") or \"200\") \n",
    "bootstrap_sampled_fraction = float(input(\"Fraction of subarray sampled for each bootstrap run? If nothing is entered, it is set to 0.8. If -1 is entered, bootstrapping is turned off\") or \"-1\")                   \n",
    "n_bootstrap_runs = int(input(\"Number of bootstrap runs? If nothing is entered, it is set to 200.\") or \"200\") \n",
    "CI_percent = float(input(\"Condidence interval? If nothing is entered, it is set to 95.\") or \"95\")\n",
    "\n",
    "start_time = time.time() # to start measuring runtime\n",
    "p_xi_given_zj = np.genfromtxt('different-7-peps-scores.csv', delimiter=',') # import full dataset\n",
    "inference_mode = \"peptide\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "2de71e38-1433-4f10-9f43-b1c2952e5b2d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Zj values from every bootstrap run (columns: peptides, rows: , displayed as fractions: \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.490464</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.218894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.103665</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.098993</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.074479</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>0.002105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>0.011401</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          0\n",
       "0  0.490464\n",
       "1  0.218894\n",
       "2  0.103665\n",
       "3  0.098993\n",
       "4  0.074479\n",
       "5  0.002105\n",
       "6  0.011401"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "bootstrap_EM()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e5d38b00-5014-4a8f-b69e-181b8a280f47",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "### protein inferring"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "5d6af966-e5b8-4be1-9f9d-269d893cbb0a",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# inference_mode = \"peptide\"\n",
    "# # bootstrap_EM_analytics_AVG()\n",
    "# # p_xi = np.copy(p_zj_bootstrap_results_fraction_avg, order='K', subok=False)\n",
    "# p_xi = np.array([0.2857,\n",
    "# 0.2143,\n",
    "# 0.2143,\n",
    "# 0.0714,\n",
    "# 0.0714,\n",
    "# 0.0714,\n",
    "# 0.0714\n",
    "# ])\n",
    "\n",
    "# print(\"p_xi:\")\n",
    "# display(pd.DataFrame(p_xi))\n",
    "\n",
    "# p_xi_given_zj = np.genfromtxt('different-7-peps-scores.csv', delimiter=',')\n",
    "# # print(\"p_xi_given_zj:\")\n",
    "# # display(pd.DataFrame(p_xi_given_zj))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "6adfad72-f878-41bb-a6c2-982ba82b4b09",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "#bootstrap_EM()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c8de24f2-3dbd-40be-994f-d2c4c8fd4d12",
   "metadata": {
    "tags": []
   },
   "source": [
    "## output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "da85c1a0-9a5e-4fc4-8321-c1ecdb4a4625",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AVG [%]</th>\n",
       "      <th>±STD [%]</th>\n",
       "      <th>-CI [%]</th>\n",
       "      <th>+CI [%]</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>49.0464</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>21.8894</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>10.3665</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9.8993</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>7.4479</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>0.2105</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1.1401</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   AVG [%] ±STD [%] -CI [%] +CI [%]\n",
       "0  49.0464      N/A     N/A     N/A\n",
       "1  21.8894      N/A     N/A     N/A\n",
       "2  10.3665      N/A     N/A     N/A\n",
       "3   9.8993      N/A     N/A     N/A\n",
       "4   7.4479      N/A     N/A     N/A\n",
       "5   0.2105      N/A     N/A     N/A\n",
       "6   1.1401      N/A     N/A     N/A"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[1mSettings of this run\u001b[0m\n",
      "EM_convergence_minimum: 0.0001\n",
      "EM_loopcounter_max: 200\n",
      "bootstrap_sampled_fraction: -1.0\n",
      "n_bootstrap_runs: 200\n",
      "CI_percent: 95.0 \n",
      "\n",
      "Runtime: 85.5 seconds ---\n"
     ]
    }
   ],
   "source": [
    "bootstrap_EM_analytics_AVG()\n",
    "bootstrap_CI_minmax = bootstrap_EM_analytics_CI()\n",
    "\n",
    "### STDs (alternative to CI)\n",
    "p_zj_bootstrap_results__fraction_std = np.std(p_zj_bootstrap_results_fraction, axis = 0)\n",
    "\n",
    "# Combines AVG zj values (from all bootstrap runs), the STD, and the two bounds for the user-chosen confidence interval in one  \n",
    "if bootstrap_sampled_fraction == -1: #ie if no bootstrapping. Fills -CI and +CI columns with \"N/A\" values\n",
    "    p_zj_avg_plus_CI = np.stack((p_zj_bootstrap_results_fraction_avg*100, np.full((len(p_zj_bootstrap_results_fraction_avg)), fill_value=\"N/A\", dtype=None), np.full((len(p_zj_bootstrap_results_fraction_avg)), fill_value=\"N/A\", dtype=None), np.full((len(p_zj_bootstrap_results_fraction_avg)), fill_value=\"N/A\", dtype=None)), axis=1)    \n",
    "    display(pd.DataFrame(p_zj_avg_plus_CI, columns = [\"AVG [%]\", \"±STD [%]\", \"-CI [%]\", \"+CI [%]\"]).astype({\"AVG [%]\":float}).round(4))\n",
    "else:\n",
    "    p_zj_avg_plus_CI = np.stack((p_zj_bootstrap_results_fraction_avg, p_zj_bootstrap_results__fraction_std, bootstrap_CI_minmax[0], bootstrap_CI_minmax[1]), axis = 1)\n",
    "    display(pd.DataFrame(p_zj_avg_plus_CI, columns = [\"AVG [%]\", \"±STD [%]\", \"-CI [%]\", \"+CI [%]\"]).round(4)*100)\n",
    "    \n",
    "### print user-settings\n",
    "print('\\033[1m' + 'Settings of this run' + '\\033[0m')\n",
    "print(\"EM_convergence_minimum:\", EM_convergence_minimum)\n",
    "print(\"EM_loopcounter_max:\", EM_loopcounter_max)\n",
    "print(\"bootstrap_sampled_fraction:\", bootstrap_sampled_fraction)\n",
    "print(\"n_bootstrap_runs:\", n_bootstrap_runs)\n",
    "print(\"CI_percent:\", CI_percent, \"\\n\")\n",
    "\n",
    "### print other\n",
    "print(\"Runtime: %s seconds ---\" % round((time.time() - start_time),1)) #run time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31a15c3f-77e8-40f7-8c04-a2ddc013b7a7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0dad34d2-1df7-4627-b2df-e6b5df1cfd87",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82aeeff0-1016-43f6-b7e6-a1e6d7612b4e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c0d6f685-de20-4340-ad3f-6d860d7ce53e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7190bd10-57f2-4653-aee1-6e47a1b82580",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3dbf494e-2e2b-4f40-bcfd-666892266987",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
